Twilio-org/phonebank
# [Ticket 134] Server Detects End Of Campaign

**Description:**

- Update recordAttempt in calls controller to check for last attempted call and update campaign status to 'completed'
  - Added the following helpers:
    - saveResponses - separates out saving responses so that sending status headers can be handled properly at end of campaign
    - getCallsNotAttempted - gets all calls not marked with a status of 'ATTEMPTED'
    - isFinalOutcome - checks if the outcome of the call is in a 'final state', ie 'ANSWERED', 'DO_NOT_CALL', or 'BAD_NUMBER'
    - isLastCallAttempted - checks if the call is the last one to be marked as completed, ie all calls are marked as 'ATTEMPTED' and the outcome of this call is in a final state
    - markCampaignAsCompleted - marks the campaign as completed
  - Updated the following helper:
    - afterPut - removed sending status from helper and added it into controller logic to support the last call scenario
- Add getCallsNotAttemptedByCampaignId service to calls service
- Add markCampaignAsCompleted to campaigns service
- Add tests for services above

Tests:
- run `npm run test:server`
- Manually:
  - Set all campaign calls except for 3 as status='ATTEMTPED' and outcome='BAD_NUMBER'
  - Login as a volunteer with your phone number connected to the account
  - Join the above campaign
  - Click Start Calling
    - Mark one contact as 'DO_NOT_CALL' 
    - Mark one contact as 'BAD_NUMBER'
    - Mark one contact as 'ANSWERED' and enter responses
  - Check db to make sure campaign was marked as status='completed', responses were properly saved,  and all calls were properly updated with 'ATTEMPTED' and the proper call outcome
